### PR TITLE
Add more prefetch conditions

### PR DESCRIPTION
**Pull request by afcapel**
This PR adds some more conditions where a link is not considered safe to be prefetched.

The risk/reward ratio favors a more conservative approach. If we don't prefetch a link the interaction can be a bit slower, but if we prefetch a link that is not safe, we may trigger a bug.

The conditions are quite specific, in any case, so they shouldn't trigger often.

The new conditions are:

- If the link is a UJS link (i.e. it has a `data-remote` attribute).
- If the link has a data-turbo-confirm attribute.
- If the link has a data-turbo-stream attribute.

The PR also cleans up the code a bit, extracting each of the conditions into helper methods.

Fixes https://github.com/hotwired/turbo/issues/1170
Fixes https://github.com/hotwired/turbo/issues/1174

### CODE REVIEW

File: src/observers/link_prefetch_observer.js [hunk 1]
```diff
@@ -145,50 +139,52 @@ export class LinkPrefetchObserver {
...
-      return false
-    }
+    if (!href) return false
 
-    const event = dispatch("turbo:before-prefetch", {
-      target: link,
-      cancelable: true
-    })
+    if (unfetchableLink(link)) return false
+    if (linkToTheSamePage(link)) return false
```
Comment (kevinmcconnell) [inline src/observers/link_prefetch_observer.js:145]: As we discussed IRL, I'm not sure we should always exclude the current page. Sometimes clicking navigation to the same page is a valid way of getting latest content on a page, and in that case prefetching seems valid.

Perhaps there are edge cases to think about when fetching them, though. In any case, probably fine to leave for another PR, just wanted to note is since we're touching this part.

File: src/observers/link_prefetch_observer.js [hunk 2]
```diff
@@ -145,50 +139,52 @@ export class LinkPrefetchObserver {
...
+    if (linkOptsOut(link)) return false
+    if (nonSafeLink(link)) return false
+    if (eventPrevented(link)) return false
```
Comment (kevinmcconnell) [inline src/observers/link_prefetch_observer.js:148]: This is much nicer with these checks extracted and given meaningful names, btw! 👏

**Review by kevinmcconnell — APPROVED** [review]
Looks great!

**Comment by seanpdoyle** [comment]
Should we also open PR to remove https://github.com/hotwired/turbo-site/blob/694d442e6e57208814803cb5b2536e93bb398cd6/_source/handbook/02_drive.md#L369-L374 from the examples for canceling a prefetch?

File: src/observers/link_prefetch_observer.js (hunk 1 above)
Comment (ur5us) [reply src/observers/link_prefetch_observer.js:145]: You could probably argue that if the current page is prefetched it might as well morph it to the latest version. Though there are probably edge cases as you say.

**Review by ur5us — COMMENTED** [review]
Thanks for addressing this so quickly.

**Comment by afcapel** [comment]
@seanpdoyle good point, I've removed the example in https://github.com/hotwired/turbo-site/pull/162
